sktime/datatypes/tests/test_vectorize.py

```diff
@@ -0,0 +1,321 @@
+# -*- coding: utf-8 -*-
+"""Testing vectorization via VectorizedDF."""
+
+__author__ = ["fkiraly"]
+
+import numpy as np
+import pytest
+
+from sktime.datatypes import MTYPE_REGISTER, SCITYPE_REGISTER
+from sktime.datatypes._check import check_is_mtype
+from sktime.datatypes._examples import get_examples
+from sktime.datatypes._vectorize import VectorizedDF
+from sktime.utils._testing.deep_equals import deep_equals
+
+SCITYPES = ["Panel", "Hierarchical"]
+
+
+def _get_all_mtypes_for_scitype(scitype):
+    """Return list of all mtypes for scitype.
+
+    Parameters
+    ----------
+    scitype : str - scitype
+
+    Returns
+    -------
+    mtypes : list of str - list of mtypes for scitype
+    """
+    if scitype not in [s[0] for s in SCITYPE_REGISTER]:
+        raise RuntimeError(scitype + " is not in the SCITYPE_REGISTER")
+    mtypes = [key[0] for key in MTYPE_REGISTER if key[1] == scitype]
+
+    if len(mtypes) == 0:
+        # if there are no mtypes, this must have been reached by mistake/bug
+        raise RuntimeError("no mtypes defined for scitype " + scitype)
+
+    return mtypes
+
+
+def _generate_scitype_mtype_combinations():
+    """Return scitype/mtype tuples for pytest_generate_tests.
+
+    Fixtures parameterized
+    ----------------------
+    scitype : str - scitype of fixture
+    mtype : str - mtype of fixture
+    """
+    # collect fixture tuples here
+
+    sci_mtype_tuples = []
+
+    for scitype in SCITYPES:
+
+        mtypes = _get_all_mtypes_for_scitype(scitype)
+
+        for mtype in mtypes:
+            sci_mtype_tuples += [(scitype, mtype)]
+
+    return sci_mtype_tuples
+
+
+def _generate_scitype_mtype_fixtureindex_combinations():
+    """Return fixture tuples for pytest_generate_tests.
+
+    Fixtures parameterized
+    ----------------------
+    scitype : str - scitype of fixture
+    mtype : str - mtype of fixture
+    fixture_index : int - index of fixture tuple with that scitype and mtype
+    """
+    # collect fixture tuples here
+
+    sci_mtype_tuples = _generate_scitype_mtype_combinations()
+
+    sci_mtype_index_tuples = []
+
+    for tuple_j in sci_mtype_tuples:
+        scitype = tuple_j[0]
+        mtype = tuple_j[1]
+        n_fixtures = len(get_examples(mtype=mtype, as_scitype=scitype))
+
+        for i in range(n_fixtures):
+            sci_mtype_index_tuples += [(scitype, mtype, i)]
+
+    return sci_mtype_index_tuples
+
+
+def pytest_generate_tests(metafunc):
+    """Test parameterization routine for pytest.
+
+    Fixtures parameterized
```

Comment:
Same

Reply:
`pytest_generate_tests` is a way to parameterize fixtures that cannot be easily written as cartesian products.
The fixtures that this parameterizes are listed, so people reading the code have access to the information easily.